pkg/platform/packet/packet.go

```diff
@@ -237,6 +255,79 @@ func (c *config) terraformSmartApply(ex *terraform.Executor, dnsProvider dns.DNS
 	return ex.Apply()
 }
 
+// terraformAddDependencies adds explicit dependencies to cluster nodes so nodes
+// with a specific hw reservation UUID are created before nodes that don't have
+// a specific hw reservation UUID.
+// The function modifies c.NodesDependsOn and c.workerPools[].NodesDependsOn,
+// assigning to them a string, formatted as a terraform list, of all worker
+// terraform targets needed to be created first. For example:
+//
+//	// Suppose worker pool "example" is the only to use a specific hw
+//	// reservation IDs. IOW, the controllers (c.NodesDependsOn) depends on
+//	// worker pool "example" to be created first.
+//	// Then, after calling this function, the attribute will be:
+// 	c.NodesDependsOn = "[ module.worker-example.worker_nodes_ids ]"
+//
+//
+// The explicit terraform dependency is needed to guarantees that nodes using
+// hardware reservation "next-available" won't use reservation IDS that another
+// worker pool may specify with a specific UUID, and thus fail to create the
+// node. This race condition is best explained here, if you want more info:
+// https://github.com/terraform-providers/terraform-provider-packet/issues/176
+// https://github.com/terraform-providers/terraform-provider-packet/pull/208
+func (c *config) terraformAddDependencies() {
+	// Nodes with specific hw reservation IDs
+	nodesWithRes := make([]string, 0)
+
+	if len(c.ReservationIDs) > 0 {
+		// Use a dummy tf output to wait on controllers nodes
+		tfTarget := clusterTarget(c.ClusterName, "controller_nodes_ids")
+		nodesWithRes = append(nodesWithRes, tfTarget)
+	}
+
+	for _, w := range c.WorkerPools {
+		if len(w.ReservationIDs) > 0 {
+			// Use a dummy tf output to wait on workers nodes
+			tfTarget := poolTarget(w.Name, "worker_nodes_ids")
+			nodesWithRes = append(nodesWithRes, tfTarget)
+		}
+	}
+
+	// Collected all nodes with reservations, create a dependency on others
+	// to them, so those nodes are created first.
+
+	// Convert nodesWithRes slice to a terraform-formatted list.
+	tfDeps := fmt.Sprintf("[ %v ]", strings.Join(nodesWithRes, ", "))
+
+	if len(c.ReservationIDs) == 0 {
+		c.NodesDependsOn = tfDeps
+	}
+
+	for i := range c.WorkerPools {
+		if len(c.WorkerPools[i].ReservationIDs) > 0 {
+			continue
+		}
+
+		c.WorkerPools[i].NodesDependsOn = tfDeps
+	}
+}
+
+// poolToTarget returns a string that can be used as "-target" argument to terraform.
+// For example:
+//	// target will be "module.worker-pool1.ex".
+//	target := poolTarget("pool1", "ex")
+func poolTarget(name, resource string) string {
+	return fmt.Sprintf("module.worker-%v.%v", name, resource)
+}
+
+// clusterTarget returns a string that can be used as "-target" argument to terraform.
+// For example:
+//	// target will be "module.packet-clusterName.ex".
+//	target := clusterTarget("clusterName", "ex")
+func clusterTarget(name, resource string) string {
+	return fmt.Sprintf("module.packet-%v.%v", name, resource)
```

Comment:
We should get rid of that and simply name controller module `module.controller` or something, as it doesn't need to be unique in `lokoctl`.

Reply:
I should perhaps create an issue about it.

Reply:
I think that even if we do that (probably we should), having a function to go from a cluster in go to a terraform target seems reasonable. It seems boilerplate code, otherwise (although not too much, so... I doubt a little bit) :)

Reply:
So, IMHO it is fine like this. If you think it doesn't need to be addressed here, please open an issue and tag me so we can discuss it there.

Re-open the discussion if this is not addressed (sorry, I think it is, but I might be missing something :))

Reply:
Yeah, no worries. Created an issue for that: https://github.com/kinvolk/lokomotive/issues/308.